sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs

```diff
@@ -0,0 +1,197 @@
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.FlowAnalysis.Common;
+using SonarAnalyzer.Helpers.FlowAnalysis.CSharp;
+using ExplodedGraph = SonarAnalyzer.Helpers.FlowAnalysis.CSharp.ExplodedGraph;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ObjectsShouldNotBeDisposedMoreThanOnce : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3966";
+        private const string MessageFormat = "Refactor this code to make sure Dispose is only called once on '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterExplodedGraphBasedAnalysis((e, c) => CheckForNullDereference(e, c));
+        }
+
+        private static void CheckForNullDereference(ExplodedGraph explodedGraph, SyntaxNodeAnalysisContext context)
+        {
+            var objectDisposedCheck = new ObjectDisposedPointerCheck(explodedGraph);
+            explodedGraph.AddExplodedGraphCheck(objectDisposedCheck);
+
+            var nullIdentifiers = new HashSet<IdentifierNameSyntax>();
+
+            EventHandler<ObjectDisposedEventArgs> memberAccessedHandler =
+                (sender, args) =>
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(rule, args.DisposableIdentifier.GetLocation(), args.DisposableIdentifier.GetText()));
+                };
+
+            objectDisposedCheck.ObjectDisposed += memberAccessedHandler;
+
+            try
+            {
+                explodedGraph.Walk();
+            }
+            finally
+            {
+                objectDisposedCheck.ObjectDisposed -= memberAccessedHandler;
+            }
+        }
+
+        internal class ObjectDisposedEventArgs : EventArgs
+        {
+            public SyntaxNode DisposableIdentifier { get; }
+
+            public ObjectDisposedEventArgs(SyntaxNode disposableIdentifier)
+            {
+                DisposableIdentifier = disposableIdentifier;
+            }
+        }
+
+        internal sealed class ObjectDisposedPointerCheck : ExplodedGraphCheck
+        {
+            public event EventHandler<ObjectDisposedEventArgs> ObjectDisposed;
+
+            public ObjectDisposedPointerCheck(ExplodedGraph explodedGraph)
+                : base(explodedGraph)
+            {
+            }
+
+            public override ProgramState PreProcessInstruction(ProgramPoint programPoint, ProgramState programState)
+            {
+                var instruction = programPoint.Block.Instructions[programPoint.Offset];
+                switch (instruction.Kind())

```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace this 'switch' statement with 'if' statements to increase readability. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1301)